### PR TITLE
MergeBlocks: Rewrite to use a generic algorithm

### PR DESCRIPTION
**Pull request by kripken**
Before this we had special logic for various call types. This replaces all
that with a single general code path, which unifies everything except
for control flow constructs (which remain as before, handled in a
special way for each of them).

The algorithm is simple and direct, basically it goes through the
children and when it finds a block, it sees if it can move the block's
contents outside of the parent. While doing so it takes into account
effects and so forth.

To make this easy, a random-access API is added to ChildIterator.

Diff without whitespace makes the existing test updates a lot simpler.
Note that this is not NFC as the old algorithm had some quirks like
not taking into account effects when there were more than 2 children;
the new code is uniform in how it handles things.

This ends up removing 19% of all blocks in j2wasm, which reduces
1% of total code size.

### CODE REVIEW

File: src/passes/ReorderFunctions.cpp
Comment (tlively) [inline src/passes/ReorderFunctions.cpp:28]: Intentional drive-by?

**Review by tlively — APPROVED** [review]
LGTM modulo those questions.

File: src/passes/MergeBlocks.cpp
Comment (kripken) [reply src/passes/MergeBlocks.cpp:581]: When replacing something with something else of a different type we'd need to run refinalize to update all the types etc. It's simpler to leave that to other passes to do beforehand. I'll add a comment.

File: src/passes/MergeBlocks.cpp
Comment (kripken) [reply src/passes/MergeBlocks.cpp:602]: I considered that too. I'm not sure it's more efficient, as to add the effects we end up doing a merge operation. How much that saves later might make it worthwhile, but I'm not sure. (Perhaps we should add an API to scan and merge as a single operation?)

File: src/passes/ReorderFunctions.cpp
Comment (kripken) [reply src/passes/ReorderFunctions.cpp:28]: Oh, sorry, yes, I guess I had this in my tree. I can remove it. I meanwhile opened #4322 anyhow.

**Pull request merged**
